instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/TracingExecutionInterceptor.java

```diff
@@ -125,35 +126,22 @@ public SdkRequest modifyRequest(
     return request;
   }
 
+  @NoMuzzle
```

Comment:
`@NoMuzzle` turns off a pretty important security feature of the agent -- in general I think I'd prefer all the no-muzzle'd methods to be kept as minimal as possible, contained in classes separate from the rest of the code, and either guarded with extra defensive conditions (e.g. it'd be great to move/copy the `SqsSendMessageRequestAccess.isInstance` check to the beginning of this method) or at the very least heavily commented why it is safe.

With that in mind, WDYT about moving all the affected code into a separate class, and minimizing the entry points? It'd be great if it was just one method that was used outside of it, and the rest of them would be private and encapsulated in the new class.

Reply:
If you do this then you may run into an issue where classes used only in the `@NoMuzzle` method are not found. This is because besides the safety checks muzzle plugin also composes the list of the instrumentation classes used. You may need to manually add these classes using `getAdditionalHelperClassNames` method as described in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md#inject-additional-instrumentation-helper-classes-manually-with-the-getadditionalhelperclassnames-method

Reply:
Hmm, that's true, although it's true for the current code as well. E.g. `MessageAttributeValueMapGetter` is only referenced in a `@NoMuzzle` method.

If we want to avoid that, we could modify the `ReferenceCollectingClassVisitor` to create a special `MethodVisitor` for `@NoMuzzle` methods, that just collects helper classes and not references.

Reply:
> If you do this then you may run into an issue where classes used only in the @NoMuzzle method are not found. 

I don't think that's true, it seems that `@NoMuzzle` is not transitive. I.e. if a method annotated with NoMuzzle calls another method, that method is checked again -- I even had to make a lambda a separate method to add NoMuzzle.

So it unfortunately won't help us minimize NoMuzzle if we minimize the entrypoints. But it would still probably be a nice refactoring to move out SQS-specific code to a separate class.

> @NoMuzzle turns off a pretty important security feature of the agent

But: If you use reflection instead, you turn off pretty important safety features of the Java compiler (I think you mean safety here, not security). And IIUC, reflection stops Muzzle from noticing problems just like `@NoMuzzle` (though it is more fine grained with reflection)

Reply:
> So it unfortunately won't help us minimize NoMuzzle if we minimize the entrypoints. But it would still probably be a nice refactoring to move out SQS-specific code to a separate class.

👍 

> It will not be that well-encapsulated, since we need a few different hooks (I think two) where we do SQS specific things.

That's fine. I just want to make sure that any usage of `@NoMuzzle` is very deliberate and well documented.

> But: If you use reflection instead, you turn off pretty important safety features of the Java compiler (I think you mean safety here, not security).

Yup, safety, I mixed up these two for some reason 🙈 
That's true; that's why we've had all these null checks on `MethodHandle`s.

Reply:
> > If you do this then you may run into an issue where classes used only in the @NoMuzzle method are not found.
> 
> I don't think that's true, it seems that `@NoMuzzle` is not transitive. I.e. if a method annotated with NoMuzzle calls another method, that method is checked again -- I even had to make a lambda a separate method to add NoMuzzle.
> 
I suspect that it will behave differently when you are calling a mehtod in a class that isn't used anywhere else from `@NoMuzzle`. The reason is that muzzle runs on the whole class, once a class is discovered all methods will be checked and new classes will be added to the list of dependencies. `@NoMuzzle` will prevent discovering the classes that are used only inside the `@NoMuzzle` method.

Reply:
@laurit What will this mean then? If I "lose" a class due to this, I would get a NoClassDefFound error at runtime because it would not be injected in the right class loader?

Reply:
> @laurit What will this mean then? If I "lose" a class due to this, I would get a NoClassDefFound error at runtime because it would not be injected in the right class loader?

yes, that is what I would expect

Reply:
I wonder if it would be worth adding some muzzle enhancement like `@NoMuzzle(trackDependencies=true)`. But I don't know how involved that would be, it might not be too hard given that all the code is in this repository.

Reply:
It seemed so trivial and simple when I first had the idea 😅 Now I'm starting to wonder if I should continue the reflection way after all, but I will take a look at your suggestions. Thank you both!

Reply:
Since you demonstrated that this actually works, I integrated your multi-module approach from https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8603#issuecomment-1578491584 (cherry-picked as b941bad).

To me it looks like we now arrived at a sustainable pattern where we don't work too much against Muzzle but rather with it. Do you agree? If so, I can would expand this to replace the remaining reflection-based bridge classes.

Reply:
Oh well, actually it doesn't work 😞 

```
 > Task :instrumentation:aws-sdk:aws-sdk-2.2:javaagent:muzzle-AssertPass-software.amazon.awssdk-aws-core-2.2.0 FAILED
FAILED MUZZLE VALIDATION: io.opentelemetry.javaagent.instrumentation.awssdk.v2_2.SqsInstrumentationModule mismatches:
-- io.opentelemetry.javaagent.shaded.instrumentation.awssdk.v2_2.SqsImpl:28 Missing class software.amazon.awssdk.services.sqs.model.SendMessageRequest
-- io.opentelemetry.javaagent.shaded.instrumentation.awssdk.v2_2.SqsImpl:45 Missing class software.amazon.awssdk.services.sqs.model.MessageAttributeValue$Builder
-- io.opentelemetry.javaagent.shaded.instrumentation.awssdk.v2_2.SqsImpl:39 Missing class software.amazon.awssdk.services.sqs.model.SendMessageRequest$Builder
-- io.opentelemetry.javaagent.shaded.instrumentation.awssdk.v2_2.SqsImpl:45 Missing class software.amazon.awssdk.services.sqs.model.MessageAttributeValue
```

I wonder why that is, does this task assume just a single muzzled module?

Reply:
I think I found a good solution to this build failure by testing the core module and the sqs module in separate muzzle runs using excludedInstrumentationNames. When you have multiple other modules (e.g. sqs, and sns) it could get a bit unwieldy, but then adding a feature includedInstrumentationNames to muzzle might not be that hard.

Reply:
I think `excludeInstrumentationName` is a very good solution for now. I didn't know we had such an option, would have thought that you'd need to use `extraDependency` to get around this.

Reply:
Removed the whole agent-unit-tests module in a095c37. We now lose the check that we actually do reference SQS from the SQS module, but that sanity check does not seem valuable enough to justify the module.

Reply:
Do you think this GH conversation can be considered "resolved" now?